lib/pry/plugins.rb

```diff
@@ -35,6 +35,7 @@ def enable!
       # Load the Command line options defined by this plugin (if they exist)
       def load_cli_options
         cli_options_file = File.join(spec.full_gem_path, "lib/#{spec.name}/cli.rb")
+        cli_options_file = File.realpath(cli_options_file) if RUBY_VERSION >= "2.4.4"
```

Comment:
This check is not very accurate. Could we use `Gem::Version.new` to make this comparison?

Reply:
This is a good question, in fact the first version I sent did just that, but I changed my mind, because I was not sure about several things:
- do we always assume that `Gem` is available? I know it is usually bundled with ruby, but maybe we need to use pry to debug `rubygems`.
- do we require `gem` (usually there is no need since pry will be loaded by rubygems) here?
- Gem::Version can raise exceptions (malformed version number string), do we catch them?

So I thought about it, and realized that we probably don't need to do a Gem::Version comparison immediately. Indeed, the check will be correct as long as the "major.minor" version of ruby, major or minor are not greater than 10.

But I agree it is brittle, so if you agree with my original version (which used Gem::Version.new instead), I'll rebase on master with this version.

Reply:
Good thinking, however we already use this API in this codebase so we're not introducing a new dependency: https://github.com/pry/pry/search?q=Gem%3A%3AVersion&unscoped_q=Gem%3A%3AVersion

> do we always assume that Gem is available?

I think as of now it's OK to add this. If we want to be able to debug Rubygems itself, we should probably resolve this separately (probably by vendoring `Gem::Version`).

> do we require gem 

I believe you mean `'rubygems'`? That used to be an 1.8 practice but with modern Rubies it's not needed anymore and Pry doesn't support 1.8 anymore.

> Gem::Version can raise exceptions

That's a good point. I believe the exception will be raised when you load Pry, so it doesn't really matter that much because the problem will show itself immediately, right?

Like you said, I find it brittle, too. Who knows how people version their gems, so it's better to rely on APIs that handle that IMO.

Reply:
> Good thinking, however we already use this API in this codebase so we're not introducing a new dependency: https://github.com/pry/pry/search?q=Gem%3A%3AVersion&unscoped_q=Gem%3A%3AVersion

Ok, great, so let's use that.

> > do we always assume that Gem is available?
 
> I believe you mean `'rubygems'`? That used to be an 1.8 practice but with modern Rubies it's not needed anymore and Pry doesn't support 1.8 anymore.

I was wondering about more exotic rubies like `mruby`, I don't know if `rubygems` is always automatically loaded here (but I never tested it).
 
> That's a good point. I believe the exception will be raised when you load Pry, so it doesn't really matter that much because the problem will show itself immediately, right?

Yes, the exception would be raised when loading plugins, so at the start. I still added a rescue just in case.
 
> Like you said, I find it brittle, too. Who knows how people version their gems, so it's better to rely on APIs that handle that IMO.

But this only apply to RUBY_VERSION which is fairly straightforward to parse. But I agree with you, let's not overcomplicate thing and just use Gem::Version.

So I did that and rebased onto master.

Reply:
Pry does not support mruby, at least explicitly. I think it’s fine not to `rescue` there because we have never claimed that it does. Other rubies we support are known to have `Gem` from the beginning. The return value from rescue (`true`) doesn’t seem to make sense either. I would not overcomplicate until we hit real issues.

Reply:
The `rescue` is not there to protect against a non existing `Gem`, but to protect against Gem::Version(RUBY_VERSION) giving an argument error. I return true because I catch the exception inside the if, and in this case I assume we have a newer ruby (with a strange version string).

I can stll remove the rescue if you don't like it...

Reply:
That would be my preference 😛

Reply:
Done